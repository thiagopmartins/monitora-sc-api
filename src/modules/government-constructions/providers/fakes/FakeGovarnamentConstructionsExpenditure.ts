import GovernmentConstructionsExpenditure from '@modules/government-constructions/infra/typeorm/schemas/GovernmentConstructionsExpenditure';
import AppError from '@shared/errors/AppError';
import IGovernmentConstructionsExpenditureProvider from '../IGovernmentConstructionsExpenditureProvider';

class FakeGovarnamentConstructionsExpenditure
  implements IGovernmentConstructionsExpenditureProvider {
  public async getAll(): Promise<
    GovernmentConstructionsExpenditure[] | undefined
  > {
    try {
      const construction = {
        infos: [
          {
            statusId: 3,
            statusDesc: 'a iniciar',
            projectsAmount: 46,
            projectsValue: 6.287068161e7,
          },
          {
            statusId: 2,
            statusDesc: 'em andamento',
            projectsAmount: 185,
            projectsValue: 2.56298338198e9,
          },
          {
            statusId: 1,
            statusDesc: 'conclu√≠das',
            projectsAmount: 2370,
            projectsValue: 4.30347594188e9,
          },
          {
            statusId: 4,
            statusDesc: 'paralisadas',
            projectsAmount: 62,
            projectsValue: 1.24280871189e9,
          },
          {
            statusId: 5,
            statusDesc: 'rescindidas',
            projectsAmount: 35,
            projectsValue: 3.575703148e7,
          },
        ],
        urlInfoPanel:
          'http://www.sicop.sc.gov.br:80/relatoriosExterno/abrirRelSomatorioObras.do',
      };

      return construction.infos as GovernmentConstructionsExpenditure[];
    } catch (error) {
      throw new AppError(error);
    }
  }
}

export default FakeGovarnamentConstructionsExpenditure;
