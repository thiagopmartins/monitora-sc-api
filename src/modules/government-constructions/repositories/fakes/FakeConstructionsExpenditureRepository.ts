import ICreateConstructionsExpenditureDTO from '@modules/government-constructions/dtos/ICreateConstructionsExpenditureDTO';
import GovernmentConstructionsExpenditure from '@modules/government-constructions/infra/typeorm/schemas/GovernmentConstructionsExpenditure';
import { uuid } from 'uuidv4';
import IGovernmentConstructionsExpenditureRepository from '../IGovernmentConstructionsExpenditureRepository';

class FakeConstructionsExpenditureRepository
  implements IGovernmentConstructionsExpenditureRepository {
  public expenditures: GovernmentConstructionsExpenditure[] = [];

  public async create(
    expenditureDTO: ICreateConstructionsExpenditureDTO,
  ): Promise<GovernmentConstructionsExpenditure> {
    const expenditure = new GovernmentConstructionsExpenditure();

    const date = new Date();

    Object.assign(
      expenditure,
      { id: uuid(), created_at: date, updated_at: date },
      expenditureDTO,
    );
    this.expenditures.push(expenditure);

    return expenditure;
  }

  public async findByConstructionStatusId(
    statusId: number,
  ): Promise<GovernmentConstructionsExpenditure | undefined> {
    const findExpenditure = this.expenditures.find(
      expenditure => expenditure.statusId === statusId,
    );
    return findExpenditure;
  }

  public async update(
    data: GovernmentConstructionsExpenditure,
  ): Promise<GovernmentConstructionsExpenditure> {
    const findIndex = this.expenditures.findIndex(
      findExpenditure => findExpenditure.id === data.id,
    );

    this.expenditures[findIndex] = data;

    return data;
  }
}

export default FakeConstructionsExpenditureRepository;
