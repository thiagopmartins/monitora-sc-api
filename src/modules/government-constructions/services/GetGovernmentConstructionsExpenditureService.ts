import { inject, injectable } from 'tsyringe';
import ICreateConstructionsExpenditureDTO from '../dtos/ICreateConstructionsExpenditureDTO';
import IGovernmentConstructionsExpenditureProvider from '../providers/IGovernmentConstructionsExpenditureProvider';
import IGovernmentConstructionsExpenditureRepository from '../repositories/IGovernmentConstructionsExpenditureRepository';

@injectable()
class GetGovernmentConstructionsExpenditureService {
  constructor(
    @inject('GovernmentConstructionsExpenditureRepository')
    private governmentConstructionsExpenditureRepository: IGovernmentConstructionsExpenditureRepository,

    @inject('GovernmentConstructionsExpenditureProvider')
    private governmentConstructionsExpenditureProvider: IGovernmentConstructionsExpenditureProvider,
  ) {}

  public async execute(): Promise<void> {
    const result = await this.governmentConstructionsExpenditureProvider.getAll();
    if (result !== undefined) {
      result.map(async expenditure => {
        const expenditureAlreadyExists = await this.governmentConstructionsExpenditureRepository.findByConstructionStatusId(
          expenditure.statusId,
        );

        if (!expenditureAlreadyExists) {
          await this.governmentConstructionsExpenditureRepository.create(
            expenditure as ICreateConstructionsExpenditureDTO,
          );
        } else {
          await this.governmentConstructionsExpenditureRepository.update(
            expenditureAlreadyExists,
          );
        }
      });
    }
  }
}

export default GetGovernmentConstructionsExpenditureService;
