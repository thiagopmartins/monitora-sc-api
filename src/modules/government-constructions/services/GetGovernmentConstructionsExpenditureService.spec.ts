import FakeGovarnamentConstructionsExpenditure from '../providers/fakes/FakeGovarnamentConstructionsExpenditure';
import FakeConstructionsExpenditureRepository from '../repositories/fakes/FakeConstructionsExpenditureRepository';
import GetGovernmentConstructionsExpenditureService from './GetGovernmentConstructionsExpenditureService';

describe('GetGovernmentConstructionsExpenditure', () => {
  let fakeConstructionsExpenditureRepository: FakeConstructionsExpenditureRepository;
  let fakeGovarnamentConstructionsExpenditure: FakeGovarnamentConstructionsExpenditure;
  let getGovernmentConstructionsExpenditure: GetGovernmentConstructionsExpenditureService;

  beforeAll(() => {
    fakeConstructionsExpenditureRepository = new FakeConstructionsExpenditureRepository();
    fakeGovarnamentConstructionsExpenditure = new FakeGovarnamentConstructionsExpenditure();
    getGovernmentConstructionsExpenditure = new GetGovernmentConstructionsExpenditureService(
      fakeConstructionsExpenditureRepository,
      fakeGovarnamentConstructionsExpenditure,
    );
  });

  it('should be able to create a new construction expenditure', async () => {
    const create = jest.spyOn(fakeConstructionsExpenditureRepository, 'create');

    await getGovernmentConstructionsExpenditure.execute();

    expect(create).toBeCalledWith(
      expect.objectContaining({
        statusId: expect.any(Number),
      }),
    );

    fakeConstructionsExpenditureRepository.expenditures.forEach(expenditure => {
      expect(expenditure).toMatchSnapshot({
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
        id: expect.any(String),
      });
    });
  });

  it('should be able to update if already exists construcion expenditure', async () => {
    await getGovernmentConstructionsExpenditure.execute();

    const update = jest.spyOn(fakeConstructionsExpenditureRepository, 'update');

    await getGovernmentConstructionsExpenditure.execute();
    await getGovernmentConstructionsExpenditure.execute();

    expect(update).toBeCalledWith(
      expect.objectContaining({
        statusId: expect.any(Number),
      }),
    );

    fakeConstructionsExpenditureRepository.expenditures.forEach(expenditure => {
      expect(expenditure).toMatchSnapshot({
        created_at: expect.any(Date),
        updated_at: expect.any(Date),
        id: expect.any(String),
      });
    });
  });

  it('should be not call repository functions if governmentConstructionsAmountProvider is undefined', async () => {
    jest
      .spyOn(fakeGovarnamentConstructionsExpenditure, 'getAll')
      .mockImplementation(() => Promise.resolve(undefined));

    const find = jest.spyOn(
      fakeConstructionsExpenditureRepository,
      'findByConstructionStatusId',
    );

    await getGovernmentConstructionsExpenditure.execute();

    expect(find).not.toBeCalled();
  });
});
