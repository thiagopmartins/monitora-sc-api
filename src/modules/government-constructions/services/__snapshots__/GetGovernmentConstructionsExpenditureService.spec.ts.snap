// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetGovernmentConstructionsExpenditure should be able to create a new construction expenditure 1`] = `
Object {
  "created_at": Any<Date>,
  "id": Any<String>,
  "projectsAmount": 46,
  "projectsValue": 62870681.61,
  "statusDesc": "a iniciar",
  "statusId": 3,
  "updated_at": Any<Date>,
}
`;

exports[`GetGovernmentConstructionsExpenditure should be able to create a new construction expenditure 2`] = `
Object {
  "created_at": Any<Date>,
  "id": Any<String>,
  "projectsAmount": 185,
  "projectsValue": 2562983381.98,
  "statusDesc": "em andamento",
  "statusId": 2,
  "updated_at": Any<Date>,
}
`;

exports[`GetGovernmentConstructionsExpenditure should be able to create a new construction expenditure 3`] = `
Object {
  "created_at": Any<Date>,
  "id": Any<String>,
  "projectsAmount": 2370,
  "projectsValue": 4303475941.88,
  "statusDesc": "concluídas",
  "statusId": 1,
  "updated_at": Any<Date>,
}
`;

exports[`GetGovernmentConstructionsExpenditure should be able to create a new construction expenditure 4`] = `
Object {
  "created_at": Any<Date>,
  "id": Any<String>,
  "projectsAmount": 62,
  "projectsValue": 1242808711.89,
  "statusDesc": "paralisadas",
  "statusId": 4,
  "updated_at": Any<Date>,
}
`;

exports[`GetGovernmentConstructionsExpenditure should be able to create a new construction expenditure 5`] = `
Object {
  "created_at": Any<Date>,
  "id": Any<String>,
  "projectsAmount": 35,
  "projectsValue": 35757031.48,
  "statusDesc": "rescindidas",
  "statusId": 5,
  "updated_at": Any<Date>,
}
`;

exports[`GetGovernmentConstructionsExpenditure should be able to update if already exists construcion expenditure 1`] = `
Object {
  "created_at": Any<Date>,
  "id": Any<String>,
  "projectsAmount": 46,
  "projectsValue": 62870681.61,
  "statusDesc": "a iniciar",
  "statusId": 3,
  "updated_at": Any<Date>,
}
`;

exports[`GetGovernmentConstructionsExpenditure should be able to update if already exists construcion expenditure 2`] = `
Object {
  "created_at": Any<Date>,
  "id": Any<String>,
  "projectsAmount": 185,
  "projectsValue": 2562983381.98,
  "statusDesc": "em andamento",
  "statusId": 2,
  "updated_at": Any<Date>,
}
`;

exports[`GetGovernmentConstructionsExpenditure should be able to update if already exists construcion expenditure 3`] = `
Object {
  "created_at": Any<Date>,
  "id": Any<String>,
  "projectsAmount": 2370,
  "projectsValue": 4303475941.88,
  "statusDesc": "concluídas",
  "statusId": 1,
  "updated_at": Any<Date>,
}
`;

exports[`GetGovernmentConstructionsExpenditure should be able to update if already exists construcion expenditure 4`] = `
Object {
  "created_at": Any<Date>,
  "id": Any<String>,
  "projectsAmount": 62,
  "projectsValue": 1242808711.89,
  "statusDesc": "paralisadas",
  "statusId": 4,
  "updated_at": Any<Date>,
}
`;

exports[`GetGovernmentConstructionsExpenditure should be able to update if already exists construcion expenditure 5`] = `
Object {
  "created_at": Any<Date>,
  "id": Any<String>,
  "projectsAmount": 35,
  "projectsValue": 35757031.48,
  "statusDesc": "rescindidas",
  "statusId": 5,
  "updated_at": Any<Date>,
}
`;
